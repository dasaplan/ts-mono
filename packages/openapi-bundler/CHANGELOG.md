# @dasaplan/openapi-bundler

## 0.0.7

### Patch Changes

- ef2415e: fixed toposort schemas when array items are allOf objects

## 0.0.6

### Patch Changes

- fix mergeAlOf order to follow default behaviour

## 0.0.5

### Patch Changes

- fix mergeAllOf order to follow default behaviour

## 0.0.3

### Patch Changes

- Update documentation

## 0.0.2

### Patch Changes

- handle dangling properties for schemas with allOf array.
