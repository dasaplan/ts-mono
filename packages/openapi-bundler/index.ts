export {
  bundleOpenapi,
  createCommandBundle,
  createSpecProcessor,
  cleanObj,
  OpenApiBundled,
  parseOpenapi,
  Resolver,
  Transpiler,
  OaComponent,
  SchemaResolverContext,
  Schema,
} from "./src/index.js";
