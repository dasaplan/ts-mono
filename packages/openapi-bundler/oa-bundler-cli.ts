#!/usr/bin/env ts-node

import { Command } from "commander";
import * as process from "process";
import { createCommandBundle } from "./src/index.js";

const program = new Command();
createCommandBundle(program);

program.parse(process.argv);
