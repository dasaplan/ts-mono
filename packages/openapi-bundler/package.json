{
  "name": "@dasaplan/openapi-bundler",
  "description": "Opinionated bundler/merger for Openapi (Swagger)",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "version": "0.0.0",
  "keywords": [
    "openapi",
    "typescript",
    "ts",
    "zod",
    "codegen"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/dasaplan/ts-mono.git",
    "directory": "packages/openapi-bundler"
  },
  "scripts": {
    "prepack": "sort-package-json --quiet && clean-package",
    "postpack": "clean-package restore",
    "build": "pnpm rup",
    "rup": "tsup-node --dts",
    "lint": "eslint \"./src/*.ts\" index.ts oa-bundler-cli.ts",
    "clean": "rimraf dist",
    "test": "pnpm test:esm",
    "test:esm": "cross-env NODE_OPTIONS=--experimental-vm-modules jest",
    "test:esm:update": "cross-env NODE_OPTIONS=--experimental-vm-modules jest -u"
  },
  "bin": {
    "openapi-bundler-cli": "dist/oa-cli.js"
  },
  "files": ["dist"],
  "dependencies": {
    "commander": "^11.1.0",
    "json-schema-merge-allof": "^0.8.1",
    "jsonpointer": "^5.0.1",
    "openapi3-ts": "^4.3.3",
    "merge-json-schemas": "^1.0.0",
    "tslog": "^4.9.2",
    "lodash": "^4.17.21",
    "toposort": "^2.0.2"
  },
  "peerDependencies": {
    "@redocly/openapi-core": "^1.6.0"
  },
  "devDependencies": {
    "tsconfig": "workspace:*",
    "@types/json-schema-merge-allof": "^0.6.5",
    "@types/lodash":"^4.17.5",
    "@dasaplan/ts-sdk": "workspace:*"
  },
  "clean-package": {
    "remove": [
      "scripts",
      "devDependencies",
      "tsup",
      "clean-package",
      "dependencies.@dasaplan/ts-sdk"
    ]
  },
  "tsup": {
    "entry": [
      "index.ts",
      "oa-bundler-cli.ts"
    ],
    "noExternal": [
      "@dasaplan/ts-sdk",
      "tsconfig"
    ],
    "format": [
      "cjs",
      "esm"
    ],
    "tsconfig": "tsconfig.build.json",
    "clean": true,
    "sourcemap": true
  }
}
