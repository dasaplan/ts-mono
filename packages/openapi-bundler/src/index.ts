export { bundleOpenapi, OpenApiBundled, parseOpenapi } from "./bundle.js";
export { createCommandBundle } from "./commands.js";
export { createSpecProcessor } from "./post-process/index.js";
export { Transpiler, Schema } from "./transpiler/index.js";
export {
  Resolver,
  OaComponent,
  SchemaResolverContext,
  cleanObj,
} from "./resolver/index.js";
