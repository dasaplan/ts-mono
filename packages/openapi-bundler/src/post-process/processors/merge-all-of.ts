/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/no-unused-vars */

import { OpenApiBundled } from "../../bundle.js";
import { oas30 } from "openapi3-ts";

import jsonSchemaMergeAllOff from "json-schema-merge-allof";
import { isRef } from "@redocly/openapi-core";
import { appLog } from "../../logger.js";
import { _, ApplicationError } from "@dasaplan/ts-sdk";
import { cleanObj, SchemaResolverContext } from "../../resolver/index.js";
import { mergeXOmit, XOmitConfig } from "./x-omit-deep.js";

export function mergeAllOf(bundled: OpenApiBundled) {
  const log = appLog.childLog(mergeAllOf);

  const mergedAllOf = _.cloneDeep(bundled);
  const { collected, ctx } = findSchemaObjectsWithAllOf(mergedAllOf);
  log.info(`mergeAllOf. Merging allOf arrays in ${collected.length} schemas`);
  log.debug(`mergeAllOf. Merging in schemas: ${collected.map((s) => s.id).join(", ")}`);

  collected.forEach((s) => {
    try {
      doMerge(s, ctx);
    } catch (e) {
      throw ApplicationError.create(`failed to merge allOf for schema.id ${s.id}`).chainUnknown(e);
    }
  });
  return mergedAllOf;
}

function doMerge({ schema, id }: { id: string; schema: any }, ctx: SchemaResolverContext) {
  const log = appLog.childLog(doMerge);
  const subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject> = schema.allOf ?? [];
  // include dangling properties {allOf: [], danglingA: {}, danglingB: [], danglingC: null, ... }
  const danglingProperties = { ..._.omit(schema, "allOf") };
  if (!_.isEmpty(danglingProperties)) {
    log.debug(`danglingProperties. fixing dangling properties for schema.id ${id} - moving all properties into allOf array`);
    schema.allOf?.push(danglingProperties);
    cleanObj(schema, ["allOf"]);
  }
  // when we merge allOf array, we do not want to mutate referenced entities
  // omitting clone will mutate "schema"
  const clonedSchemas = _.cloneDeep(subSchemas);
  // omitting clone will mutate cached references
  const resolvedSchemas = _.cloneDeep(resolveSubSchemas(clonedSchemas, ctx));
  if (!_.isEmpty(danglingProperties)) {
    resolvedSchemas.push({ pointer: undefined, resolved: danglingProperties });
  }
  const hasOneOfOrAnyOf = resolvedSchemas.filter((f) => !_.isEmpty(f.resolved.oneOf) || !_.isEmpty(f.resolved.anyOf))?.length > 0;
  if (hasOneOfOrAnyOf) {
    throw ApplicationError.create(`unsupported allOf element. anyOf or oneOf are yet not supported as an allOf subschema: ${JSON.stringify(schema)}`);
  }

  // scenario(real inheritance): a allOf.element holds a ref where the target has a discriminator
  const parentsWithDiscriminator = resolvedSchemas.filter((p) => _.isDefined(p.pointer) && _.isDefined(p.resolved.discriminator));

  // scenario(composition): a allOf.element is a schema
  const resolvedSchemasWithoutDiscriminator = resolvedSchemas.filter((p) => !_.isDefined(p.resolved.discriminator));

  // scenario(composition of a base schema): a allOf.element defines a schema with a discriminator by reference and inlined
  const inlinesWithDiscriminator = resolvedSchemas.filter((p) => !_.isDefined(p.pointer) && _.isDefined(p.resolved.discriminator));
  if (parentsWithDiscriminator.length == 1 && inlinesWithDiscriminator.length > 0) {
    // we may have an allOf expressing inheritance and multiple schemas we need to merge (while keeping inherited schema)
    const mergedWithInlineParent = mergeSubSchemas([...inlinesWithDiscriminator, ...resolvedSchemasWithoutDiscriminator], ctx);
    // build hierarchy
    schema.allOf = [{ $ref: parentsWithDiscriminator[0].pointer }, mergedWithInlineParent.resolved];
    return schema;
  }
  if (resolvedSchemas.length < 1) {
    // something is off: could be a schema with a single allOf or an allOf comprised of multiple discriminators...
    log.warn(`found allOf with single element in schema.id '${id}': ${JSON.stringify(schema)}`);
    return schema;
  }

  if (parentsWithDiscriminator.length == 1 && resolvedSchemasWithoutDiscriminator.length == 1) {
    // we may have an allOf expressing inheritance - most tooling can handle two allOfs
    return schema;
  }

  if (parentsWithDiscriminator.length == 1 && resolvedSchemasWithoutDiscriminator.length > 1) {
    // we may have an allOf expressing inheritance and multiple schemas we need to merge (while keeping inherited schema)
    const merged = mergeSubSchemas(resolvedSchemasWithoutDiscriminator, ctx);
    // build hierarchy
    schema.allOf = [{ $ref: parentsWithDiscriminator[0].pointer }, merged.resolved];
    return schema;
  }

  if (parentsWithDiscriminator.length > 1) {
    // we may have an allOf expressing multiple inheritance we will try to find a common base

    const merged = mergeSubSchemas(resolvedSchemasWithoutDiscriminator, ctx);
    // build hierarchy
    const hierarchy = parentsWithDiscriminator.reduce((acc, curr) => {
      return { allOf: [{ $ref: curr.pointer }, acc] };
    }, merged.resolved);
    delete schema["$ref"];
    Object.assign(schema, hierarchy);
    return schema;
  }

  if (parentsWithDiscriminator.length < 1) {
    // we only have schemas which we can merge together
    const merged = mergeSubSchemas(resolvedSchemasWithoutDiscriminator, ctx);
    Object.assign(schema, merged.resolved);
    delete schema["allOf"];
    delete schema["$ref"];
    return schema;
  }

  throw "Error: should not have been reached";
}

function resolveSubSchemas(
  subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject>,
  ctx: SchemaResolverContext
): Array<{
  pointer: string | undefined;
  resolved: oas30.SchemaObject & { $ref?: string };
}> {
  const subs = subSchemas.map((d) => {
    return resolveRefNode(d, ctx, { deleteRef: true });
  });
  const allOfs = subs.flatMap((s) => s.resolved.allOf ?? []);
  if (_.isEmpty(allOfs)) {
    return subs;
  }
  const flatten = resolveSubSchemas(allOfs, ctx);
  return [...flatten, ...subs].map((s) => ({
    pointer: s.pointer,
    resolved: _.omit(s.resolved, "allOf"),
  }));
}

function getJsonSchemaMergeAllOff<T extends oas30.SchemaObject>(
  subschemas: Array<{
    pointer: string | undefined;
    resolved: T;
  }>
): T {
  try {
    return jsonSchemaMergeAllOff<oas30.SchemaObject & { "x-omit"?: XOmitConfig }>(
      { allOf: subschemas.map((s) => s.resolved) },
      {
        resolvers: {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-expect-error
          additionalProperties: (values, path, mergeSchemas, options) => {
            return jsonSchemaMergeAllOff.options.resolvers.additionalProperties(values, path, mergeSchemas, options);
          },
          // overwrite title by higher indexed schema: [ {title: a}, {title:b}] => {title:b}
          // default title means take first occurance
          defaultResolver: (values, path, mergeSchemas, options) => {
            if (path?.[0] === "x-omit") {
              const [a, b] = values;
              return mergeXOmit(a, b);
            }

            if (options.resolvers?.title) {
              // use configured resolver
              return options.resolvers.title(values, path, mergeSchemas, options);
            }
            // use fallback resolver
            return jsonSchemaMergeAllOff.options.resolvers.title(values, path, mergeSchemas, options);
          },
        },
      }
    ) as T;
  } catch (e: unknown) {
    throw ApplicationError.create("failed merging allOf sub schemas").chainUnknown(e);
  }
}

export function mergeSubSchemas(
  _resolvedSchemas: Array<{
    pointer: string | undefined;
    resolved: oas30.SchemaObject;
  }>,
  ctx: SchemaResolverContext
) {
  const subschemas = _resolvedSchemas.map((d) => ({
    pointer: d.pointer,
    resolved: ctx.resolver.resolveRefImmutable(d.resolved, { deleteRef: true }),
  }));

  const merged = getJsonSchemaMergeAllOff(subschemas.reverse());
  // clean references in properties
  Object.values(merged.properties ?? {}).forEach((propValue) => {
    if (!isRef(propValue) || _.isEmpty(_.omit(propValue, "$ref"))) {
      // property is either a ref or an object => that is fine
      return;
    }
    // we have both a ref and an object. We try to keep the ref, if the content is identical
    const schema = {
      pointer: undefined,
      resolved: _.cloneDeep(_.omit(propValue, "$ref")),
    };
    const referenced = resolveRefNode(propValue, ctx, { deleteRef: false });
    if (_.isEqual(schema.resolved, referenced.resolved)) {
      cleanObj(propValue, []);
      Object.assign(propValue, { $ref: referenced.pointer });
      return;
    }
    Object.assign(propValue, getJsonSchemaMergeAllOff([schema, referenced]));
    delete (propValue as never)["$ref"];
  });
  return {
    pointer: undefined,
    resolved: merged,
  };
}

function findSchemaObjectsWithAllOf(bundled: OpenApiBundled) {
  const resolver = SchemaResolverContext.create(bundled);
  const collected = resolver.schemas.filter((s) => _.isDefined(s.schema.allOf));
  return { collected, ctx: resolver };
}

function resolveRefNode(data: { $ref: string } | unknown, ctx: SchemaResolverContext, params?: { deleteRef: boolean }) {
  if (!isRef(data)) {
    return {
      pointer: undefined,
      resolved: ctx.resolver.resolveRefImmutable(data as oas30.SchemaObject, params),
    };
  }
  return {
    pointer: data["$ref"],
    resolved: ctx.resolver.resolveRefImmutable(data, params),
  };
}
