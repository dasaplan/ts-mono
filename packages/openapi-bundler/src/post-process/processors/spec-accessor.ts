import { OpenApiBundled } from "../../bundle.js";
import { Resolver } from "../../resolver/index.js";
import { ApplicationError } from "@dasaplan/ts-sdk";

export namespace OpenapiApiDoc {
  type ApiDoc = OpenApiBundled;

  export function accessor(bundled: ApiDoc) {
    const resolver = Resolver.create(bundled);
    return {
      getSchemaByName: createGetSchemaByName(bundled, resolver),
      getSchemaByRef: createGetSchemaByRef(bundled, resolver),
      get schemas() {
        return Object.entries(bundled.components?.schemas ?? {}).map(([name, schema]) => ({ name, schema }));
      },
    };
  }

  function createGetSchemaByName(oa: ApiDoc, resolver: Resolver) {
    return (name: string) => {
      const schema = oa.components?.schemas?.[name];
      ApplicationError.assert(schema, `could not find schema.name in spec: ${name} `);
      return resolver.resolveRef(schema, { deleteRef: true });
    };
  }

  function createGetSchemaByRef(oa: ApiDoc, resolver: Resolver) {
    return (ref: string | { $ref: string }) => {
      const refObj = typeof ref === "string" ? { $ref: ref } : ref;
      const schema = resolver.resolveRef(refObj);
      ApplicationError.assert(schema, `could not find schema.$ref in spec`);
      return schema;
    };
  }
}
