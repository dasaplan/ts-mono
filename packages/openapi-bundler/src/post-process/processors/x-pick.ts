/* eslint-disable @typescript-eslint/no-explicit-any,@typescript-eslint/no-unused-vars */

import { OpenApiBundled } from "../../bundle.js";
import { oas30 } from "openapi3-ts";

import { _, ApplicationError } from "@dasaplan/ts-sdk";
import { cleanObj, SchemaResolverContext } from "../../resolver/index.js";
import { isRef } from "@redocly/openapi-core";
import { appLog } from "../../logger.js";
import { mergeSubSchemas } from "./merge-all-of.js";

export type XPickConfig = {
  required?: Array<string> | boolean;
  properties?: Record<string, boolean | XPickConfig> | boolean;
} & {
  [prop in Exclude<keyof oas30.SchemaObject, "required" | "properties">]?: boolean;
};

type PickCtx = SchemaResolverContext & { maxRecursionDepths: number };

type WithPick<T> = T & { "x-pick": Partial<XPickConfig> };
export function xPick(bundled: OpenApiBundled) {
  const log = appLog.childLog(xPick);

  const picked = _.cloneDeep(bundled);
  const { collected, ctx } = findSchemaObjectsWithXPick(picked);
  log.info(`xPick. Picking fields in ${collected.length} schemas`);
  log.debug(`xPick. Picking in schemas: ${collected.map((s) => s.id).join(", ")}`);
  collected.forEach((s, idx, all) => {
    try {
      doPick(s, { ...ctx, maxRecursionDepths: 10 });
    } catch (e) {
      throw ApplicationError.create(`failed to pick fields for schema.id ${s.id}`).chainUnknown(e);
    }
  });
  return picked;
}

function doPick({ schema, id }: { id: string; schema: any }, ctx: PickCtx) {
  if (_.isEmpty(schema.allOf)) {
    // resolve x-pick in schema: after mergeAllOf it could be the case that there is no allOf anymore
    const picked = applyPick({ id, schema, merged: schema }, ctx);
    delete (picked as any)["x-pick"];
    Object.assign(cleanObj(schema), picked);
    return;
  }

  // resolve x-pick in allOf array
  const subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject> = schema.allOf ?? [];
  const resolvedSchemas = _.cloneDeep(resolveSubSchemas(_.cloneDeep(subSchemas), ctx));

  const merged = mergeSubSchemas(resolvedSchemas, ctx)?.resolved;
  const picked = applyPick({ id, schema, merged }, ctx);
  delete (picked as any)["x-pick"];
  Object.assign(cleanObj(schema), picked);
}

function applyPick<T>(args: { id: string; schema: T; merged: WithPick<T> }, ctx: PickCtx) {
  const log = appLog.childLog(applyPick);

  const merged: oas30.SchemaObject = args.merged;
  const pickConfig: XPickConfig = merged["x-pick"];
  if (_.isNil(pickConfig)) {
    return;
  }
  const picked: oas30.SchemaObject = _.cloneDeep(merged);

  // pick required
  if (typeof pickConfig.required === "boolean") {
    // nothing to do - we will take the whole required array
  } else if (!_.isEmpty(pickConfig.required)) {
    const requiredToPick = pickConfig.required;
    picked.required = picked.required?.filter((e) => requiredToPick?.includes(e));
    if (merged.required?.length === picked.required) {
      log.warn(
        `Nothing to pick. The required array of schema.id ${args.id} does not include any defined values to pick. required(pick): ${requiredToPick?.join(
          ","
        )}  `
      );
    }
  }

  // pick properties
  if (typeof pickConfig.properties === "boolean") {
    // nothing to do - we will take the whole
  } else if (!_.isEmpty(pickConfig.properties)) {
    const propsToPick = pickConfig.properties;
    picked.properties = Object.entries(picked.properties ?? {}).reduce((acc, [key, val]) => {
      const toPick = propsToPick[key];
      if (_.isDefined(toPick) && toPick) {
        // filter in entry to pick
        return { ...acc, [key]: val };
      }
      return acc;
    }, {});
  }

  // pick all other fields
  const configWithoutPropsRequired = _.omit(pickConfig, "properties", "required");

  return Object.entries(picked).reduce((acc, [key, val]) => {
    const toPick = configWithoutPropsRequired[key as keyof typeof configWithoutPropsRequired];
    if (_.isDefined(toPick) && toPick) {
      // filter in entry to remove
      return { ...acc, [key]: val };
    }
    return acc;
  }, _.pick(picked, "properties", "required"));
}

function resolveSubSchemas(
  subSchemas: Array<oas30.ReferenceObject | oas30.SchemaObject>,
  ctx: SchemaResolverContext
): Array<{
  pointer: string | undefined;
  resolved: oas30.SchemaObject & { $ref?: string };
}> {
  const subs = subSchemas.map((d) => {
    return resolveRefNode(d, ctx, { deleteRef: true });
  });
  const allOfs = subs.flatMap((s) => s.resolved.allOf ?? []);
  if (_.isEmpty(allOfs)) {
    return subs;
  }
  const flatten = resolveSubSchemas(allOfs, ctx);
  return [...flatten, ...subs].map((s) => ({
    pointer: s.pointer,
    resolved: _.omit(s.resolved, "allOf"),
  }));
}

function resolveRefNode(data: { $ref: string } | unknown, ctx: SchemaResolverContext, params?: { deleteRef: boolean }) {
  if (!isRef(data)) {
    return {
      pointer: undefined,
      resolved: ctx.resolver.resolveRefImmutable(data as oas30.SchemaObject, params),
    };
  }
  return {
    pointer: data["$ref"],
    resolved: ctx.resolver.resolveRefImmutable(data, params),
  };
}

function findSchemaObjectsWithXPick(bundled: OpenApiBundled) {
  const resolver = SchemaResolverContext.create(bundled);
  const collected = resolver.schemas.filter(
    (s) => _.isDefined(s.schema["x-pick"]) || s.schema.allOf?.some((e) => (isRef(e) ? false : _.isDefined(e["x-pick"])))
  );
  return { collected, ctx: resolver };
}

export function mergeXPick(a: Partial<XPickConfig>, b: Partial<XPickConfig>) {
  let merged = { ...a, ...b };

  if (typeof a.required === "boolean" || typeof b.required === "boolean") {
    const pickRequiered = typeof a.required === "boolean" ? a.required : typeof b.required === "boolean" ? b.required : false;
    merged = { ...merged, required: pickRequiered };
  } else if (a.required || b.required) {
    merged = { ...merged, required: a?.required ?? b?.required };
  }

  if (typeof a.properties === "boolean" || typeof b.properties === "boolean") {
    const pickProps = typeof a.properties === "boolean" ? a.properties : typeof b.properties === "boolean" ? b.properties : false;
    merged = { ...merged, properties: pickProps };
  } else if (a.properties || b.properties) {
    merged = { ...merged, properties: a?.properties ?? b?.properties };
  }

  return merged;
}
