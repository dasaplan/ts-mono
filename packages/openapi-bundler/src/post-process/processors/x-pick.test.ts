/* eslint-disable @typescript-eslint/no-unused-vars */
import { beforeAll, describe, expect, test } from "vitest";
import { mergeAllOf } from "./merge-all-of.js";
import { BundleMock } from "@dasaplan/openapi-bundler";

import { OpenapiApiDoc } from "./spec-accessor.js";
import { appLog } from "../../logger.js";
import { xPick } from "./x-pick.js";

describe("x-pick", () => {
  beforeAll(() => {
    appLog.setLogLevel("silly");
  });
  const {
    createApi,
    withSchema,
    factory: { schemaRef, mockXPick },
  } = BundleMock.create();

  test("pick", () => {
    const spec = createApi(
      withSchema("A", {
        required: ["a"],
        properties: { a: { type: "string" } },
      }),
      withSchema("B", {
        allOf: [
          { required: ["b", "bb"], properties: { b: { type: "string" }, bb: { type: "string" } } },
          mockXPick({
            required: ["bb"],
            properties: { bb: true },
          }),
        ],
      }),
      withSchema("AB", {
        allOf: [
          schemaRef("A"),
          schemaRef("B"),
          mockXPick({
            required: ["a"],
            properties: { b: true },
          }),
        ],
      })
    );

    const picked = xPick(mergeAllOf(spec));
    const omittedAccessor = OpenapiApiDoc.accessor(picked);
    const AB_afterPick = omittedAccessor.getSchemaByName("AB");
    expect(AB_afterPick["x-pick"]).toBeUndefined();
    expect(AB_afterPick.allOf).toBeUndefined();
    expect(AB_afterPick.required).toEqual(["a"]);
    expect(AB_afterPick.properties?.a).toBeUndefined();
    expect(AB_afterPick.properties?.b).toBeDefined();
    expect(picked).toMatchSnapshot();
  });

  test("nothing to pick", () => {
    const spec = createApi(
      withSchema("B", {
        allOf: [schemaRef("Pick"), { required: ["b", "bb"], properties: { b: { type: "string" }, bb: { type: "string" } } }],
      }),
      withSchema("Pick", {
        "x-pick": { required: ["c"], properties: { c: true } },
      })
    );

    const picked = xPick(mergeAllOf(spec));
    const accessor = OpenapiApiDoc.accessor(picked);
    const schema = accessor.getSchemaByName("B");

    expect(schema, "expected schema to be defined").toBeDefined();
    expect(schema.required, "expected 'required' not to be processed").toEqual([]);
    expect(schema.properties?.b, "expected property 'b' not to be processed").toBeUndefined();
    expect(schema.properties?.bb, "expected property 'bb' not to be processed").toBeUndefined();
    expect(schema.properties?.c, "expected property 'c' not to be processed").toBeUndefined();

    expect(schema["x-pick"], "expected x-pick toBeUndefined").toBeUndefined();
  });
});
