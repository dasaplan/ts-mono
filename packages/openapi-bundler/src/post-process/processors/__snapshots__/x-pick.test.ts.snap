// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`x-pick > pick 1`] = `
{
  "components": {
    "schemas": {
      "A": {
        "properties": {
          "a": {
            "type": "string",
          },
        },
        "required": [
          "a",
        ],
        "type": "object",
      },
      "AB": {
        "properties": {
          "b": {
            "type": "string",
          },
        },
        "required": [
          "a",
        ],
      },
      "B": {
        "properties": {
          "bb": {
            "type": "string",
          },
        },
        "required": [
          "bb",
        ],
      },
    },
  },
  "info": {
    "title": "",
    "version": "",
  },
  "openapi": "3.0.3",
  "paths": {},
}
`;
